MFA login

alrocar opened the pull request:
Closes #14336

alrocar commented:
retest this please

alrocar commented:
Some screenshots:

## Individual user

- MFA that needs setup

![screen shot 2018-11-07 at 12 42 24](https://user-images.githubusercontent.com/623650/48129874-1c6c6f00-e28b-11e8-9fbe-9a134b25814c.png)
![screen shot 2018-11-07 at 12 42 31](https://user-images.githubusercontent.com/623650/48129884-20988c80-e28b-11e8-9e68-b34f01556b7f.png)

- Next times once MFA is setup

![screen shot 2018-11-07 at 12 50 51](https://user-images.githubusercontent.com/623650/48130104-cfd56380-e28b-11e8-8be0-1b88a2f03864.png)


## Organization owner

- MFA enabled that needs setup

![screen shot 2018-11-07 at 12 44 40](https://user-images.githubusercontent.com/623650/48129931-53428500-e28b-11e8-9013-b85e6cdce15b.png)
![screen shot 2018-11-07 at 12 44 45](https://user-images.githubusercontent.com/623650/48129932-563d7580-e28b-11e8-8534-c919e09f9c3b.png)

- Next times once MFA is setup

![screen shot 2018-11-07 at 12 44 18](https://user-images.githubusercontent.com/623650/48130121-da8ff880-e28b-11e8-9a61-a99b575cf3c9.png)

## Organization user

Same as organization owner, without the `skip this step` footer

## Validations

- To many invalid verification codes

![screen shot 2018-11-07 at 12 43 05](https://user-images.githubusercontent.com/623650/48129900-31490280-e28b-11e8-9311-3b7000102241.png)

- Invalid verification code

![screen shot 2018-11-07 at 12 42 49](https://user-images.githubusercontent.com/623650/48129896-2d1ce500-e28b-11e8-8be7-5e0c87d87c44.png)

- More than two minutes of inactivity

![screen shot 2018-11-07 at 12 55 31](https://user-images.githubusercontent.com/623650/48130284-6efa5b00-e28c-11e8-8c2e-8e79e064ac46.png)

javitonino commented on config/initializers/warden.rb:
```diff
@@ -268,6 +281,14 @@ def authenticate!
+    auth.session(opts[:scope])[:multifactor_authentication_last_activity] = Time.now.to_i
+  end
+
+  skip_multifactor_authentication = auth.winning_strategy && auth.winning_strategy.skip_multifactor_authentication?
```
Uhm, let's review all strategies and whether they should be affected by MFA:

User/password-like
- Password. The classic user + password login. Clearly YES.
- OAuth. Login via Google/GitHub. I'd say YES
- LDAP. User/password against an external LDAP directory service. I'd say YES.

API Key-like
- AuthApi & AnyAuthApi. Authentication via API Key. Cleary NO.
- APIAuthentication. Legacy OAuth (CARTO as a provider). Equivalent to API Key, so NO.

Signup related
- UserCreation. This is used to login the user after a new account creation (organization signup). This depends a bit on how we implement signup into an organization with mandatory MFA. But I'd say that YES, it should be affected by MFA. It's probably not going to matter since it won't be setup on first login, but just in case.
- EnableAccountToken. Same as above. The difference is this one is used when signing up into an organization that requires email validation via an activation link.

SSO
- HTTPHeaderAuthentication. SSO integration via trusted HTTP headers. I think it's not needed. But in this case, the user can just not configure MFA. If we say YES, they can optionally have MFA for this, which maybe some users want.
- SAML. Same as above, so YES

Very open for discussion, what do you think?

alrocar replied:
Initially we talked about password + OAuth.

About the rest:

- From a product perspective:

Has anyone requested to use MFA with LDAP, SAML, SSO, etc? I understand those are configured in on premises environments, where I'm not sure if anyone wants to add an additional layer of security.

Even for OAuth it might be a bit weird. Imagine I have enabled 2FA in Google and I want to login in CARTO. I have to login in Google + verification code for Google + verification code for CARTO...

About the `Signup related` strategies I don't understand what do you mean. Do you mean MFA might be enabled but signup don't require a verification code?

- From a technical perspective:

I'd have to take a look and see if LDAP and SSO strategies follow the same workflow also how to test each one of them properly.

Do you have any hint about this? 

One of my fears is incurring in higher maintenance costs for something no one uses. Specially when sometimes is not easy to test against an LDAP or similar servers.

In general, unless implementation of MFA in other strategies is absolutely trivial, I'd try to go with the bare minimum and in this case support more strategies on demand.

javitonino replied:
> Has anyone requested to use MFA with LDAP, SAML, SSO, etc?

No, but they probably will after we document that this is a feature :)

> Even for OAuth it might be a bit weird

Yes, but you can just not enable it. It's the user decision.

> About the Signup related strategies

These are only used for the first login, which is automatically after signup without you needing to enter the password. Right now, there is no way for a newly created user to be MFA enabled, so it's irrelevant. However, in the future we might want to allow for users to have MFA enabled during signup. I don't see the problem in making this strategy MFA enabled.

> I'd have to take a look and see if LDAP and SSO strategies follow the same workflow also how to test each one of them properly.

It's documented in our internal docs and it's not trivial to do. LDAP follows the user/password flow very closely (it just validates user/pass against an external database instead of the internal one). SSO is a bit more involved.

> In general, unless implementation of MFA in other strategies is absolutely trivial, I'd try to go with the bare minimum and in this case support more strategies on demand.

I think the point is that it should be absolutely trivial, just switching that flag. If everything is ideal, MFA should be completely independent of the login strategy used, so it should not matter that we enable it for all.

My preferred option would be to disable MFA if the strategy has `def store?; return false; end`. That is a warden built-in that means that the strategy does not save a session, i.e: it must pass the credential in each request. We can't enable MFA (you will need to do a MFA for each request), so it doesn't make sense. But for all other requests, the flow is the same, you do some stuff and finally call `warden.authenticate` to invoke the strategy, check everything, and create a session cookie.

If we decide not to support MFA for all strategies, it might be super-weird for users to be able to see the toggle to enable MFA if they are logging with an unsupported strategy, so we would need to make more code to determine which strategy is going to be used to see if we should display the MFA toggle or not. It makes this more complicated.

TL;DR:
I see value in supporting MFA for all strategies. It should work out of the box if MFA is independent of the login strategy, which is highly desired from a code architecture point of view (and looking at the code, it seems like it mostly is).

alrocar replied:
Yep, I see the value as well. Let me try how it'd work.

alrocar replied:
I made some manual tests with OAuth and it works well.

Besides that I added some tests for LDAP and SAML, just check that all the previous tests work for a user with MFA and add a new one to check that the redirection to `multifactor_authentication` is taking place.

My only concern is how to properly test this. Are we planning to do manual acceptance for each case? Should I dig on writing proper integration tests?

Anyway, I guess we can go with the next round of CR.

javitonino commented on app/controllers/sessions_controller.rb:
```diff
@@ -92,17 +101,53 @@ def show
+      end
+
+      user.active_multifactor_authentication.verify!(params[:code])
+      warden.session(user.username)[:multifactor_authentication_required] = false
```
Putting the comment here but it's something global. First, let me write what I understood:

1. Upon login, we check if the user has MFA enabled. If yes, we set a session variable `multifactor_authentication_required` to true
2. After MFA auth is successful, we set the variable to false
3. We check in a global controller if the flag is set to true in order to forbid access

Some potential problems:

If we forgot to set the session variable, the MFA login is skipped directly. This is something that we would need to do in central (enable this flag after login) so we are coupling central to cartodb (they need to have the same logic to set the flag).

Basically, we are making the default state (no session variable) to be the unsecure one (already logged in). I think we should flip the logic:
1. After login, we do nothing
2. After MFA auth, we set a variable to true, e.g: `multifactor_authentication_performed` 
3. In the global controller, we check `user.multifactor_auth_required? && !session[:mfa_performed]`

The main difference is that if we forget to set the session variable, the user is sent to the login screen. Also, in central we don't need to do anything after login, just redirect to cartodb which will take care of MFA. It's a bit less coupled.

I think this is generally better all-around, so surely I must be missing some cons. Let's discuss!

alrocar replied:
I see your point and makes sense to me. I made a quick change and it works!

![](https://media.giphy.com/media/11cTB6upfMOT5u/giphy.gif)

javitonino replied:
Like a charm :sparkles:

javitonino commented on app/controllers/sessions_controller.rb:
```diff
@@ -92,17 +101,53 @@ def show
+      user.reset_password_rate_limit
+    end
+
+    redirect_to url
```
This is the branch taken in case of `skip`. As far as I can see, it is only redirecting to the new URL, without touching anything in the models/session. I think it would redirect to the dashboard and then redirect back to this screen, wouldn't it?

alrocar replied:
I may have missed something about the `SKIP` flow. I understood it just skips MFA, isn't that right?

alrocar replied:
Oh I see, nvm

javitonino replied:
We have to make sure `skip` is only an option during `needs_setup?` phase. If MFA is enabled, this shouldn't be an option (not in the UI, but the controller should also verify it)

alrocar replied:
Oh OK, missed that then 👍

javitonino commented on app/controllers/sessions_controller.rb:
```diff
@@ -229,6 +281,16 @@ def github_config
 
   private
 
+  def user_inactive(user)
```
This method name is confusing to the max! :trollface: 

Maybe something like `mfa_inactivity_period_expired` or something like that? Not great either, but at least it tells me that it is related to MFA and to some inactivity period. 

`user_inactive` sounds like user is not active, as in, user is locked xD

javitonino commented on app/models/carto/user_multifactor_auth.rb:
```diff
@@ -28,6 +28,10 @@ def disabled?
       !enabled
     end
 
+    def needs_setup?
```
You can also use aliases: `alias_method :needs_setup?, :disabled?`

alrocar commented on spec/requests/sessions_controller_spec.rb:
```diff
@@ -435,6 +435,29 @@ def create_admin_user(organization)
     end
+
+    describe 'user with MFA' do
+      it_behaves_like 'SAML'
```
I wanted to test somehow an strategy with `skip_multifactor_authentication`

alrocar commented on spec/factories/user_multifactor_auths.rb:
```diff
@@ -1,5 +1,21 @@
+    end
+
+    factory :totp_needs_setup, traits: [:needs_setup]
+    factory :totp_active, traits: [:active]
```
👍 didn't know about that

alrocar commented on spec/factories/user_multifactor_auths.rb:
```diff
@@ -1,5 +1,21 @@
+      enabled true
+    end
+
+    trait :disabled do
```
I added the traits so it's easier to update the tests when we add the `status` attribute (instead of the `enabled` boolean attribute).

javitonino replied:
:+1: although maybe we don't need that field after all. Looking at this PR it's solved like:
- No MFA model -> disabled
- MFA model enabled = false -> needs_setup
- MFA model enabled = true -> enabled

and it looks like it works OK (from the CR view)

alrocar replied:
Yes, my doubt was the "reset MFA" feature.

We can do it by removing the current `user_multifactor_auth` and creating a new one (or reseting the shared_secret + change to `enabled = false`.

alrocar commented:
retest this please

javitonino commented:
My main concerns have been addressed :sparkles:. I'll leave the final pass after those changes to @gonzaloriestra to avoid too much noise.

alrocar commented on app/controllers/sessions_controller.rb:
```diff
@@ -281,7 +282,12 @@ def github_config
 
-  def user_inactive(user)
+  def set_last_mfa_activity
+    user = ::User.where(id: params[:user_id]).first || current_viewer
```
Depending on the action we are sending the `user_id` as a parameter in the form

alrocar commented on spec/requests/sessions_controller_spec.rb:
```diff
@@ -175,6 +183,28 @@
+        post create_session_url(user_domain: user_domain, email: admin_user_username, password: admin_user_password)
+
+        response.status.should == 302
+        (response.location =~ /^http\:\/\/#{admin_user_username}(.*)\/dashboard\/$/).to_i.should eq 0
```
👍 I'm leaving it as it is though.

gonzaloriestra commented:
I've been testing this and I've found some issues: 

- Timeout doesn't seem to work
- When the code attempts limit is reached, it shows a 500 error
- When MFA is enabled and configured, after the first step (user + password) you have access to some screens in `/dashboard`
- Wen MFA is enabled but not configured, logout doesn't work and it redirects to MFA configuration screen
- In the MFA configuration screen, it says: `If you have any problem with the QR code please, try with a manual code`. There's a missing comma before `please`, and I think the following could be removed. So I'd change it to: `If you have any problem with the QR code, please try with a manual code`.
- The size of the QR image is increased by CSS and it looks a bit blurry. Could it be possible to generate a bigger image instead of resizing?

alrocar commented:
I've fixed a couple of bugs: 

- **Timeout doesn't seem to work.**

There was a bug in the order of the sessions_controller hooks

- **When the code attempts limit is reached, it shows a 500 error**

Catching some exceptions, for the same reason.

But others couldn't reproduce. I'm not sure if we can test this in staging. I can make it work locally, but in staging the login is redirecting to central, that's why these issues you mention:

- **When MFA is enabled and configured, after the first step (user + password) you have access to some screens in `/dashboard`**

Cannot reproduce

- **When MFA is enabled but not configured, logout doesn't work and it redirects to MFA configuration screen**

Cannot reproduce

### Bonus

- In the MFA configuration screen, it says: `If you have any problem with the QR code please, try with a manual code`. There's a missing comma before `please`, and I think the following could be removed. So I'd change it to: `If you have any problem with the QR code, please try with a manual code`.

I copied the design but I've changed it as you suggested for the shake of legibility.

- The size of the QR image is increased by CSS and it looks a bit blurry. Could it be possible to generate a bigger image instead of resizing?

I've increased the size to 400px and it looks sharper.

Moving back to CR

javitonino commented:
About the non-reproducible errors, I'm 90% sure they are related to the static assets:
- Locally, dashboard assets are served by Rails which goes through all security hooks, including MFA
- In staging, they are served directly from staging, and they connect to `/api/v3/me` to check all security stuff (for example, locked state).

In this case, we probably need to do a check in frontend (similar to https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/dashboard/statics/static.js#L96) to see if the user is pending MFA and show the appropriate screens. To reproduce, you can add `skip_before_action :check_user_state` to `app/controllers/admin/visualizations_controller.rb`.

gonzaloriestra commented:
Nice, everything is working now! :+1: 

Tested:
- Regular / organizational users
- Password / OAuth users
- QR scan / manual code
- Skip button only for owners
- Several instances in Google Authenticator
- Codes expire after 60s
- Codes cannot be reused
- Timeout
- Rate limit
- Access denied to dashboard until MFA is passed
- QR sharpness and label
- Logout

gonzaloriestra commented on app/controllers/sessions_controller.rb:
```diff
@@ -17,10 +17,14 @@ class SessionsController < ApplicationController
   PASSWORD_LOCKED = 'password_locked'.freeze
+  MULTIFACTOR_AUTHENTICATION_INACTIVITY = 'multifactor_authentication_inactivity'.freeze
+
+  MAX_MULTIFACTOR_AUTHENTICATION_INACTIVITY = 120.seconds
```
I'm thinking that this time could be too short if you have to install the Google Authenticator app and so on, what about increasing it to 5 min?

alrocar replied:
Worst case, they just have to re-login. Also I assume anyone enabling MFA in CARTO should have it enabled for other services.
